### PR TITLE
MIDI playback survey 20241006

### PR DESCRIPTION
**Pull request by Icenowy**
Topic Description
-----------------

- gstreamer: specify default soundfont
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
- fluidsynth: update to 2.3.6
    - Add a patch to priortize pulse audio output driver.
    - RECOM soundfont-fluid-gm, which is the best GM soundfont we provide
    now.
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
- audacious: update to 4.4.1
    - Add a customization patch to let it default to
    /usr/share/soundfonts/default.sf2 .
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
- freepats: drop timidity cfg file
    Freepats isn't General MIDI compatible (lacks some instruments), so
    playing General MIDI compatible MIDI songs with it may lead to
    unexpected result (mainly some instruments get substituted by piano).
    Drop the timidity config file for preventing it from being used by
    TiMidity++ -- TiMidity++ can consume SF2-formatted
    General-MIDI-compatible sound patches like FluidR3_GM.
    The shipped config file is still installed, not to /etc now but to the
    installation destination of patch files, for needed users.
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
- wildmidi: update to 0.4.6
    - Recommend freepats because it's currently the only sound patches
    available in AOSC OS repository that is usable by WildMidi.
    - Explicitly specify output drivers.
    - Add a patch to default to "pulse" instead of "default" (which
    currently fails on AOSC OS) for alsa.
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
- timidity++: refactor and rebuild
    - Point to the default.sf2 alternatives link in the default config file.
    - Change the systemd service from a system one to user one, because
    current AOSC OS expect sound daemon to be run in user session.
    - Add flac to the dependency list because the binary is broken by
    libFLAC soname change (until this REL bump).
    - Add timidity PKGPROV because it's a usual name for this package too
    (even the binary shipped is called timidity instead of timidity++).
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
- soundfont-fluid-gm: add a default.sf2 alternatives link
    This is going to be the default sound font path for SF2-capable
    synthesizers shipped by AOSC OS.
    Signed-off-by: Icenowy Zheng <uwu@icenowy.me>

Package(s) Affected
-------------------

- audacious: 4.4.1
- fluidsynth: 2.3.6
- freepats: 20060219-5
- gstreamer: 1.24.7-1
- soundfont-fluid-gm: 3.1-2
- timidity++: 2.15.0-2
- wildmidi: 0.4.6

Security Update?
----------------

No

Build Order
-----------

```
#buildit soundfont-fluid-gm timidity++ wildmidi freepats audacious fluidsynth gstreamer
```

Test Build(s) Done
------------------

**Primary Architectures**

- [x] AMD64 `amd64`
- [x] AArch64 `arm64`
- [x] LoongArch 64-bit `loongarch64`

**Secondary Architectures**

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`
- [x] RISC-V 64-bit `riscv64`

### CODE REVIEW

**Comment by Icenowy** [comment]
RFC proposals:

**Comment by Icenowy** [comment]
RFC proposals:

- [x] ~Drop .MID support from FFmpeg. (It uses libmodplug's defaultly non-working and broken MIDI implementations (cannot handle GM banks definition))~ (Needs to patch FFmpeg, not going to handle this in this survey)
- [ ] Ship Merged GM/GS SF2 at http://git.icewy.top/icenowy/FluidR3_GM_GS (instead of dedicated GM and GS).
- [ ] Unpack Fluid R3 (or other GM SF2) as GUS patches for WildMidi.

**Comment by Icenowy** [comment]
Tested:
- Playing back MIDI with VLC.
- Playing back MIDI with Audacious.
- Playing back MIDI with FluidSynth CLI directly by attaching the MID file as the only parameter.
- Playing back MIDI with TiMidity++ CLI.
- Playing back MIDI with WildMidi CLI (correctness is known to be broken because of lack of GM-compliant GUS patches in AOSC OS)
- `fluidsynth.service` user service unit. (Auto-start not tested)
- `timidity.service` user service unit. (Auto-start not tested)

**Comment by Icenowy** [comment]
>     * [ ]  Unpack Fluid R3 (or other GM SF2) as GUS patches for WildMidi.

Trying to use https://github.com/psi29a/unsf to unpack FluidR3_GM.sf2, the resulting file could be correctly consumed by WildMidi.

**Comment by MingcongBai** [comment]
> RFC proposals:
> 
>     * [x]  ~Drop .MID support from FFmpeg. (It uses libmodplug's defaultly non-working and broken MIDI implementations (cannot handle GM banks definition))~ (Needs to patch FFmpeg, not going to handle this in this survey)
> 
>     * [ ]  Ship Merged GM/GS SF2 at http://git.icewy.top/icenowy/FluidR3_GM_GS (instead of dedicated GM and GS).
> 
>     * [ ]  Unpack Fluid R3 (or other GM SF2) as GUS patches for WildMidi.

I'm okay with both remaining proposals:

- So long as the merged SF2 has a documented generation routine (even if GUI is to be required).
- If something is needed and is big, it's fine - it's needed nonetheless.

File: app-multimedia/timidity++/spec [hunk 1]
```diff
@@ -2,5 +2,5 @@ VER=2.15.0
 SRCS="tbl::https://sourceforge.net/projects/timidity/files/TiMidity++/TiMidity++-$VER/TiMidity++-$VER.tar.gz"
```
Comment (MingcongBai) [inline app-multimedia/timidity++/spec]: Add `/download` per the Styling Manual.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

**Comment by MingcongBai** [comment]
> RFC proposals:
> 
>     * [x]  ~Drop .MID support from FFmpeg. (It uses libmodplug's defaultly non-working and broken MIDI implementations (cannot handle GM banks definition))~ (Needs to patch FFmpeg, not going to handle this in this survey)
> 
>     * [ ]  Ship Merged GM/GS SF2 at http://git.icewy.top/icenowy/FluidR3_GM_GS (instead of dedicated GM and GS).
> 
>     * [ ]  Unpack Fluid R3 (or other GM SF2) as GUS patches for WildMidi.

@Icenowy Indicated that these are to be handled in another topic, specifically: consider cutting out WildMidi dependency out of the three remaining dependees (GStreamer, which has three options for MIDI output; and QMMP, mpd, which remains to be investigated).

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**